### PR TITLE
Add EVP_PKEY_get_default_digest_name()

### PR DESCRIPTION
**Pull request by levitte**
This is needed to be able to sign with provider side keys.  Yet another fix of issues found by #10797 

Please read the commit messages.

### CODE REVIEW

**Comment by levitte** [comment]
Documentation pendeing

**Comment by levitte** [comment]
Now that #10778 is merged, this has been rebased and taken out of WIP

File: crypto/evp/m_sigver.c [hunk 1]
```diff
@@ -31,6 +31,7 @@ static int do_sigver_init(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
 {
     EVP_PKEY_CTX *locpctx = NULL;
     EVP_SIGNATURE *signature = NULL;
+    char locmdname[80] = "";     /* 80 chars should be enough */
```
Comment (t8m) [inline crypto/evp/m_sigver.c]: Could we have (perhaps internal) #define somewhere for this? Also below in the actual function implementation there is 100. Why this difference?

Comment (levitte) [reply crypto/evp/m_sigver.c]: I didn't remember between them...

A macro would be suitable.  How about an internal `OSSL_MAX_NAMESIZE`?

File: doc/man3/EVP_PKEY_get_default_digest_nid.pod [hunk 2]
```diff
@@ -2,31 +2,44 @@
...
+name for the public key signature operations associated with key
+I<pkey> into I<mdname>, up to at most I<mdname_sz> bytes including the
+ending NUL byte.
+
+EVP_PKEY_get_default_digest_nid() sets I<pnid> to the default message
+digest NID for the public key signature operations associated with key
+I<pkey>.  Note that some signature algorithms (i.e. Ed25519 and Ed448)
+do not use a digest during signing.  In this case I<pnid> will be set
+to NID_undef.  This function is only reliable for legacy keys, which
+is keys loaded from engines, or created with internal libcrypto
```
Comment (mattcaswell) [inline doc/man3/EVP_PKEY_get_default_digest_nid.pod]: which *are* keys?

File: doc/man3/EVP_PKEY_get_default_digest_nid.pod [hunk 3]
```diff
@@ -2,31 +2,44 @@
...
+implementations (B<EVP_PKEY_ASN1_METHOD>).
```
Comment (mattcaswell) [inline doc/man3/EVP_PKEY_get_default_digest_nid.pod]: I'm not sure I understood what you meant by "created with internal libcrypto implementations"

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/evp/m_sigver.c (hunk 1 above)
Comment (t8m) [reply crypto/evp/m_sigver.c]: That would be fine.

File: doc/man3/EVP_PKEY_get_default_digest_nid.pod (hunk 3 above)
Comment (levitte) [reply doc/man3/EVP_PKEY_get_default_digest_nid.pod]: Does the type **EVP_PKEY_ASN1_METHOD** help?  I'm all ears for a better term

Comment (mattcaswell) [reply doc/man3/EVP_PKEY_get_default_digest_nid.pod]: Oh - I see what you're getting at. Perhaps:

"This function is only reliable for legacy keys, which are keys loaded from engines, or created via EVP_PKEY_assign() or have a custom EVP_PKEY_ASN1_METHOD"

Comment (levitte) [reply doc/man3/EVP_PKEY_get_default_digest_nid.pod]: I reworded it a bit further, I hope you're pleased with the result.  Your input was very helpful to get there, thanks!

File: doc/man3/EVP_PKEY_get_default_digest_nid.pod (hunk 2 above)
Comment (levitte) [reply doc/man3/EVP_PKEY_get_default_digest_nid.pod]: Fixed

File: crypto/evp/m_sigver.c (hunk 1 above)
Comment (levitte) [reply crypto/evp/m_sigver.c]: We have those in a number of places, so I'll make a sweep and submit that in another PR.

Comment (levitte) [reply crypto/evp/m_sigver.c]: See #10928.  If that one gets a quick approval and merge, it'll be easy to rebase this one on top of that, and adapt accordingly.

**Review by mattcaswell — APPROVED** [review]

**Comment by levitte** [comment]
Merged.

ead0d2347a348f0916d6d25818d16d702f1d1156 EVP: Add EVP_PKEY_get_default_digest_name() and use it
8baa49aeac0d51504b8bcd0fd5c750c17af6fe62 Add answers for EVP_PKEY_get_default_digest_name() in RSA and DSA keymgmt

**Pull request closed without merging**